### ISSUE TITLE
Add/Expose GetSubFields() on IResolveFieldContext

### ISSUE THREAD

**Issue by Mithras** [graphql-dotnet/graphql-dotnet#3088]
## Summary

`IResolveFieldContext.SubFields` does not include fields from `GraphQLFragmentSpread` and `GraphQLInlineFragment` because `node.FieldDefinition.ResolvedType` [here](node.FieldDefinition!.ResolvedType!) is a base type (e.g. an interface) so [this](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/ExecutionStrategy.cs#L351) is always false and no fields from fragments are ever returned.
I'm not asking to change `IResolveFieldContext.SubFields` though. I just want an easy way of calling [GetSubFields()](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/ExecutionStrategy.cs#L200) with a specific type so that it returns fields from matching fragments (which it's fully capable of).

### Basic example

This is the easiest way I was able to come up with
```csharp
var executionContext = (ExecutionContext)typeof(ReadonlyResolveFieldContext)
    .GetField("_executionContext", BindingFlags.Instance | BindingFlags.NonPublic)
    .GetValue(ctx);
var subFields = (Dictionary<string, (GraphQLField field, FieldType fieldType)>)
    typeof(ExecutionStrategy).GetMethod("CollectFieldsFrom", BindingFlags.Instance | BindingFlags.NonPublic)
    .Invoke(executionContext.ExecutionStrategy, new object[] { executionContext, fragmentGraphType, ctx.FieldAst.SelectionSet, null });
```
which is using a bunch of reflection to get to the [CollectFieldsFrom()](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/ExecutionStrategy.cs#L220) which is doing exactly what I need.

### Motivation

I think it's a very common use case because in most cases you want and should query data source only for fields requested in query and today there is no easy way of extracting these fields unless I'm missing something obvious.

**Comment by Mithras** [comment]
Not exactly the same but related https://github.com/graphql-dotnet/graphql-dotnet/issues/2164

**Comment by Shane32** [comment]
Well, would I would have thought that `SubFields` would include fragments.  If we look at `SetSubFieldNodes` we see this code:

```cs
            var parentType = parent.GetObjectGraphType(context.Schema)!;
            var fields = System.Threading.Interlocked.Exchange(ref context.ReusableFields, null);
            fields = CollectFieldsFrom(context, parentType, parent.SelectionSet!, fields);
```

Perhaps the first line of code is doing something that `SubFields` isn't; that should probably be considered a bug.

Also, `IResolveFieldContext` can and possibly should be able to return the `IExecutionStrategy` in use, at which point `GetSubFields` could be called directly.  In addition, an overload to `IExecutionStrategy.GetSubFields` could be added for more convenient querying of any node without having to create an `ExecutionNode` instance.

I'd have to write some tests for further investigation.  Would you be interested in making a PR?  If so, I would first attempt to replicate the problem with a failing test.  It should just run a query and within the field resolver of the parent, it should check that SubFields returns the proper child field.  You can use a `[Theory]` to run the same test code three times - once with a normal query, once with a fragment, once with an inline fragment.  Two should fail the other pass.  Then I would attempt to fix the `SubFields` field by copying the `var parentType = parent.GetObjectGraphType(context.Schema)` line of code from the `SetSubFieldNodes` method into the `ExecutionStrategy.GetSubFields` method.  Then all 3 tests should pass (if that fixes the problem).  If it doesn't fix the problem, I can investigate further.  Once the problem is fixed, I would imagine that we would accept the PR so long as the tests are in place.  Any changes to interfaces are breaking changes and could not be released until GraphQL v6.

**Comment by Shane32** [comment]
Actually I think I misread your original question..... as an interface it can't know what fields are returned from fragment spreads of another type ...... ok..... 

(sorry I don't use interfaces or unions in my code...yet...)

**Comment by Shane32** [comment]
Well, `CollectFieldsFrom` is protected.  So you could create a derived class with a public method that calls `CollectFieldsFrom`.

```cs
public static class MyExecutionStrategyHelper
{
    private static MyExecutionStrategy _strategy = new();

    public static Dictionary<string, (GraphQLField field, FieldType fieldType)> CollectFieldsFrom(ExecutionContext context, IGraphType specificType, GraphQLSelectionSet selectionSet, Dictionary<string, (GraphQLField field, FieldType fieldType)>? fields)
        => _strategy.CollectFieldsFrom2(context, specificType, selectionSet, fields);

    private class MyExecutionStrategy : SerialExecutionStrategy //doesn't matter which
    {
        public Dictionary<string, (GraphQLField field, FieldType fieldType)> CollectFieldsFrom2(ExecutionContext context, IGraphType specificType, GraphQLSelectionSet selectionSet, Dictionary<string, (GraphQLField field, FieldType fieldType)>? fields)
            => CollectFieldsFrom(context, specificType, selectionSet, fields);
    }
}

**Comment by Shane32** [comment]
As a side note, none of the included execution strategy implementations override the method, and there's no instance members, so it makes no difference that it's not the execution strategy in use.  In a prior version of GraphQL, it was a public static method, I believe, but having everything as public static methods did not leave room for implementations to override functionality to fix bugs or alter behavior.

**Comment by Mithras** [comment]
I'll try your idea with custom class inheriting from `ExecutionStrategy`, thanks! While it's a good workaround to avoid using reflection, I still believe it's a valid feature request to simplify getting fields for interfaces.

**Comment by Shane32** [comment]
I agree. It would require adding the method to IExecutionStrategy, and the ability to grab the strategy to IResolveFieldContext.

**Comment by Mithras** [comment]
Update on my current workaround:
```csharp
private class CustomExecutionStrategy : ExecutionStrategy
{
    public override Task ExecuteNodeTreeAsync(ExecutionContext context, ExecutionNode rootNode) =>
        throw new NotImplementedException();

    public Dictionary<string, (GraphQLField field, FieldType fieldType)> GetSubFields(
        IResolveFieldContext fieldContext,
        IGraphType specificType)
    {
        var executionContext = new ExecutionContext { Schema = fieldContext.Schema, Document = fieldContext.Document };
        return CollectFieldsFrom(executionContext, specificType, fieldContext.FieldAst.SelectionSet, null);
    }
}
```
and then
```csharp
private static readonly CustomExecutionStrategy CustomExecutionStrategyInstance = new();
// ...
var subFields = CustomExecutionStrategyInstance.GetSubFields(ctx, fragmentGraphType);
```

**Comment by Shane32** [comment]
`ExecutionContext` is a rather large object.  You could probably gain performance by using `Interlocked.Exchange` to re-use a shared instance if it isn't already in use.  Actually being above to measure the gain may be debatable.

**Comment by Mithras** [comment]
Yeah, I'll probably object pool it when I'm done with everything else
